### ISSUE TITLE
Array.IndexOf option to use IComparable

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/runtime#40410]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/idea/973668/arrayindexof-option-to-use-icomparable.html)._

---
In an application I did not want to use bool Equals(Object^ b)override  since it messed up Object.Reference equals in other parts.
If I read the documentation right, Array uses equals to find a match, and with IComparable a zero would be a match. Would it be desirable to add such?

---
#### Feedback Bot on 4/1/2020, 08:20 PM: 

Thank you for taking the time to provide your suggestion.  We will do some preliminary checks to make sure we can proceed further.  We'll provide an update once the issue has been triaged by the product team.

**Comment by ygra** [comment]
The same might probably apply to `List<T>`. And perhaps `IEqualityComparer<T>` would be a better choice than an `IComparable` as only equality is needed here, not an ordering.

**Comment by GrabYourPitchforks** [comment]
I don't think we have existing `IndexOf` methods that take an equality comparer. What's the proposal here? Is it the below?

```cs
public class Array
{
    public static int IndexOf<T>(T[] array, T value, IEqualityComparer<T> equalityComparer);
    public static int IndexOf<T>(T[] array, T value, int offset, int count, IEqualityComparer<T> equalityComparer);
}
```

Would it make more sense to be a LINQ extension method?

```cs
public static class Enumerable
{
    public static int IndexOf<T>(this IEnumerable<T> enumerable, T value, IEqualityComparer<T> equalityComparer);
}
```

**Comment by svick** [comment]
Somewhat related discussion about `IndexOf` on `IEnumerable<T>`: https://github.com/dotnet/runtime/issues/31156.
